ntp-proto/src/filter.rs

```diff
@@ -0,0 +1,283 @@
+        let num_invalid_tuples = self
+            .register
+            .iter()
+            .rev()
+            .take_while(|t| t.is_dummy())
+            .count();
+
+        let num_valid_tuples = self.register.len() - num_invalid_tuples;
+
+        &self.register[..num_valid_tuples]
+    }
+
+    /// #[no_run]
+    ///                     i=n-1
+    ///                     ---     epsilon_i
+    ///      epsilon =       \     ----------
+    ///                      /        (i+1)
+    ///                     ---     2
+    ///                     i=0
+    /// Invariant: the register is sorted wrt delay
+    fn dispersion(self) -> NtpDuration {
+        self.register
+            .iter()
+            .enumerate()
+            .map(|(i, t)| t.dispersion / 2i64.pow(i as u32 + 1))
+            .fold(NtpDuration::default(), |a, b| a + b)
+    }
+
+    /// #[no_run]
+    ///                          +-----                 -----+^1/2
+    ///                          |  n-1                      |
+    ///                          |  ---                      |
+    ///                  1       |  \                     2  |
+    ///      psi   =  -------- * |  /    (theta_0-theta_j)   |
+    ///                (n-1)     |  ---                      |
+    ///                          |  j=1                      |
+    ///                          +-----                 -----+
+    ///
+    /// Invariant: the register is sorted wrt delay
+    fn jitter(self, s: &System, smallest_delay: FilterTuple) -> f64 {
+        let register = self.valid_tuples();
+
+        let root_mean_square = register
+            .iter()
+            .map(|t| (t.offset - smallest_delay.offset).to_seconds().powi(2))
+            .sum::<f64>()
+            .sqrt();
+
+        // root mean square average (RMS average). - 1 to exclude the smallest_delay
+        let jitter = root_mean_square / (register.len() - 1) as f64;
+
+        // In order to ensure consistency and avoid divide exceptions in other
+        // computations, the psi is bounded from below by the system precision
+        // s.rho expressed in seconds.
+        jitter.max(s.precision)
+    }
+}
+
+pub struct System {
+    precision: f64,
+    poll: NtpDuration,
+    leap_indicator: NtpLeapIndicator,
+}
+
+impl System {
+    #[cfg(test)]
+    fn dummy() -> Self {
+        Self {
+            precision: 0.0,
+            poll: NtpDuration::default(),
+            leap_indicator: NtpLeapIndicator::NoWarning,
+        }
+    }
+}
+
+pub struct Peer {
+    clock_filter: ClockFilterContents,
+    t: NtpTimestamp,
+
+    jitter: f64,
+    dispersion: NtpDuration,
+
+    offset: NtpDuration,
+    delay: NtpDuration,
+
+    burst_counter: u32,
+}
+
+pub struct LocalClock {
+    t: NtpTimestamp,
+}
+
+#[allow(dead_code)]
+pub fn clock_filter(
+    peer: &mut Peer,
+    s: &System,
+    c: &LocalClock,
+    clock_offset: NtpDuration,
+    roundtrip_delay: NtpDuration,
+    dispersion: NtpDuration,
+) {
```

Comment:
I don't disagree, but this still seems like the fastest way forward.

Disentangling the state is tricky. I've been continuing work on clock selection and combining, and it's obvious that you run into issues at some point, where either you need to clone or use lifetimes/RC/... It's also clear that the `struct`s in the reference implementation are way too big and monolithic. 

But, I like getting it down into code where we can have the compiler help us. To me, that is just the easiest way of making progress, rather than studying a diagram. 

There are a lot of moving parts, and we don't need to get this right on the first draft, though it is of course vital we get it right eventually.

Reply:
I disagree then with this approach, because starting from the current structure presented will also influence where we end up, and as we already know we want something radically different, I'd far rather take a starting point that is at least a little bit closer to what we eventually want, instead of this starting point where not only is it already very clear that it is not the way forward, but it also pretty much precludes any testing.